File: drills/elixir/solutions/nc95_target_sum.exs
defmodule Solution do
  # The state that matters is only the running total, not which signs produced
  # it -- so carry a map from reachable total to how many ways reach it, and
  # widen it by each number twice, once added and once subtracted. Different
  # sign choices landing on the same total merge, which is what stops the count
  # being exponential in work.
  def find_target_sum_ways(nums, target) do
    nums
    |> Enum.reduce(%{0 => 1}, fn n, totals ->
      Enum.reduce(totals, %{}, fn {total, count}, following ->
        following
        |> Map.update(total + n, count, &(&1 + count))
        |> Map.update(total - n, count, &(&1 + count))
      end)
    end)
    |> Map.get(target, 0)
  end
end

File: drills/go/solutions/nc95_target_sum.go
package main

func findTargetSumWays(nums []int, target int) int {
	// ways[s] counts the sign assignments of the prefix summing to s,
	// offset so negative sums index the slice.
	total := 0
	for _, n := range nums {
		total += n
	}
	if target > total || target < -total {
		return 0
	}
	ways := make([]int, 2*total+1)
	ways[total] = 1
	for _, n := range nums {
		next := make([]int, 2*total+1)
		for s, count := range ways {
			if count == 0 {
				continue
			}
			next[s+n] += count
			next[s-n] += count
		}
		ways = next
	}
	return ways[target+total]
}
